change fields container for Entries

kardeiz opened the pull request:
See #90. I made the current tests pass but didn't add any.

There are a few different ways to handle this, and as I note in #90, this is a breaking change, so feel free to ignore if you like.

Thanks!

abonander commented:
I think using `HashMap<Vec<String, String>>` would suffice. What do you think of making it one map for both text and file fields, since they can be handled essentially the same?

kardeiz commented:
You mean `HashMap<String, Vec<String>>`?

I think `Vec<(String, String)>` is preferable since single-valued fields are definitely the more common scenario, and putting all the single values into vecs seems like it would add some unnecessary overhead for those cases.

> What do you think of making it one map for both text and file fields, since they can be handled essentially the same?

That's totally your call, but personally I like having them separate

abonander commented:
Yeah I meant `HashMap<String, Vec<String>>`, I typed that on mobile.

We can, of course, use something like [`SmallVec`](https://docs.rs/smallvec/0.4.3/smallvec/struct.SmallVec.html), we can even use an inline array size of 1 since that's the most common case.

abonander commented:
@kardeiz are you still interested in getting this merged? you never responded

kardeiz commented:
Oh, right, sorry. I personally don't like the extra complexity and additional dependency of the `SmallVec` solution. I think the right container here is `Vec<(String, String)>`. This is going to be a breaking change whether the container is `HashMap<String, Vec<String>>` or `Vec<(String, String)>`, and I think for most uses the latter will be more convenient.

I think my PR should also implement `IntoIterator<Item=(String, String)>` for `FieldsWrapper`. This would make all of the examples work as before, without having to call `into_map` as I changed in this PR. `.iter` and `.iter_mut` will be covered by the `Deref`/`DerefMut`. We could drop the `into_inner` here, but keep the `into_map` and `into_multivalued_map` for when users would prefer to work with a `HashMap`.

Let me know what you think and I can update this PR. Thanks!

abonander commented:
With `Vec<(String, String)>` I was a little concerned about keeping extra copies of field names around but that's going to be the rarer case anyway. I think what you've proposed should be pretty good.

kardeiz commented:
I added the `IntoIterator` impl, cleaned up the examples, and added a brief note about duplicate fields. Please let me know if there is anything else. Thanks!

abonander commented on src/server/save.rs:
```diff
@@ -388,11 +388,57 @@ impl SavedFile {
+    }
+}
+
+impl ::std::iter::IntoIterator for FieldsWrapper {
```
`IntoIterator` should be in the prelude, no need to use the fully-qualified name.

kardeiz replied:
Ah, I did not know that, thanks!

abonander commented on src/server/save.rs:
```diff
@@ -388,11 +388,57 @@ impl SavedFile {
+/// enable accessing fields with duplicated names.
+
+#[derive(Debug, Default)]
+pub struct FieldsWrapper(Vec<(String, String)>);
```
WRT the name, I think `TextFields` or something would be a little clearer.

kardeiz replied:
I agree `FieldsWrapper` is not the best name. I don't know about `TextFields` though; users might then expect the files stuff to be `FileFields`/`file_fields`. Maybe `FieldsVec` or just `Fields`?

abonander commented on src/server/save.rs:
```diff
@@ -388,11 +388,57 @@ impl SavedFile {
+        &mut self.0
+    }
+}
+
```
Can you add this:

```rust
// this allows use of both `from(FieldsWrapper)` and `FieldsWrapper.into()`
impl From<FieldsWrapper> for Vec<(String, String)> {
    fn from(fields: FieldsWrapper) -> Self {
        fields.0
    }
}
```

kardeiz replied:
I've added this&mdash;

abonander commented:
Cool, just a couple nits and it'll be ready to merge I think.

abonander commented on src/server/save.rs:
```diff
@@ -388,11 +388,57 @@ impl SavedFile {
+    }
+
+    /// Converts the contents into a single-valued hashmap
+    pub fn into_map(self) -> HashMap<String, String> {
```
This seems to be strictly less useful than `into_multivalued_map`

kardeiz replied:
I've included this in case anyone was relying on the prior behavior of dropping values for duplicated field names, or if someone knows they won't have duplicate fields and wants the ease of working with a `HashMap<String, String>`.

abonander commented:
Thanks!
